File: app/src/main/java/com/wmods/wppenhacer/xposed/core/components/WaContactWpp.java
package com.wmods.wppenhacer.xposed.core.components;

import com.wmods.wppenhacer.xposed.core.devkit.Unobfuscator;
import com.wmods.wppenhacer.xposed.utils.ReflectionUtils;

import org.luckypray.dexkit.query.enums.StringMatchType;

import java.io.File;
import java.lang.reflect.Field;
import java.lang.reflect.Method;

import de.robv.android.xposed.XC_MethodHook;
import de.robv.android.xposed.XposedBridge;

public record WaContactWpp(Object mInstance) {

    public static Class<?> TYPE;

    // Fields
    private static Field fieldContactData;
    private static Field fieldUserJid;
    private static Field fieldGetWaName;

    // Methods
    private static Method getWaContactMethod;
    private static Method methodGetDisplayName;
    private static Method getProfilePhoto;

    // Instances
    private static Object mInstanceGetWaContact;
    private static Object mInstanceGetProfilePhoto;

    public WaContactWpp(Object mInstance) {
        if (TYPE == null)
            throw new RuntimeException("WaContactWpp not initialized");
        if (mInstance == null)
            throw new RuntimeException("object is null");
        if (!TYPE.isInstance(mInstance))
            throw new RuntimeException("object is not a WaContactWpp");
        this.mInstance = TYPE.cast(mInstance);
    }

    public static void initialize(ClassLoader classLoader) {
        try {
            TYPE = Unobfuscator.loadWaContactClass(classLoader);
            var classPhoneUserJid = Unobfuscator.findFirstClassUsingName(classLoader, StringMatchType.EndsWith,
                    "jid.PhoneUserJid");
            var classJid = Unobfuscator.findFirstClassUsingName(classLoader, StringMatchType.EndsWith, "jid.Jid");

            var phoneUserJid = ReflectionUtils.getFieldByExtendType(TYPE, classPhoneUserJid);
            if (phoneUserJid == null) {
                var contactDataClass = Unobfuscator.loadWaContactData(classLoader);
                fieldContactData = ReflectionUtils.getFieldByType(TYPE, contactDataClass);
                fieldUserJid = ReflectionUtils.getFieldByExtendType(contactDataClass, classJid);
            } else {
                fieldUserJid = ReflectionUtils.getFieldByExtendType(TYPE, classJid);
            }
            methodGetDisplayName = Unobfuscator.loadWaContactDisplayNameMethod(classLoader);
            fieldGetWaName = Unobfuscator.loadWaContactGetWaNameField(classLoader);

            getWaContactMethod = Unobfuscator.loadGetWaContactMethod(classLoader);
            Class<?> contactManagerClass = getWaContactMethod.getDeclaringClass();

            // Try to find existing instance via static method (getInstance pattern)
            for (Method m : contactManagerClass.getDeclaredMethods()) {
                if (java.lang.reflect.Modifier.isStatic(m.getModifiers())
                        && m.getReturnType() == contactManagerClass
                        && m.getParameterCount() == 0) {
                    try {
                        Object instance = m.invoke(null);
                        if (instance != null) {
                            mInstanceGetWaContact = instance;
                            XposedBridge.log("WAE: WaContactWpp: Captured instance via static method: " + m.getName());
                            break;
                        }
                    } catch (Exception ignored) {
                    }
                }
            }

            if (mInstanceGetWaContact == null) {
                XposedBridge.hookAllConstructors(contactManagerClass, new XC_MethodHook() {
                    @Override
                    protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                        mInstanceGetWaContact = param.thisObject;
                        XposedBridge.log("WAE: WaContactWpp: Captured instance via constructor");
                    }
                });
            } else {
                XposedBridge.log("WAE: WaContactWpp: Instance already captured, skipping constructor hook");
            }

            getProfilePhoto = Unobfuscator.loadGetProfilePhotoMethod(classLoader);

            XposedBridge.hookAllConstructors(getProfilePhoto.getDeclaringClass(), new XC_MethodHook() {
                @Override
                protected void afterHookedMethod(MethodHookParam param) throws Throwable {
                    mInstanceGetProfilePhoto = param.thisObject;
                }
            });

        } catch (Exception e) {
            XposedBridge.log(e);
        }
    }

    public Object getObject() {
        return mInstance;
    }

    public FMessageWpp.UserJid getUserJid() {
        try {
            if (fieldContactData != null) {
                var coreData = fieldContactData.get(mInstance);
                var userjid = fieldUserJid.get(coreData);
                return new FMessageWpp.UserJid(userjid);
            }
            return new FMessageWpp.UserJid(fieldUserJid.get(mInstance));
        } catch (Exception e) {
            XposedBridge.log(e);
        }
        return null;
    }

    public String getDisplayName() {
        try {
            return (String) methodGetDisplayName.invoke(mInstance);
        } catch (Exception e) {
            XposedBridge.log(e);
        }
        return null;
    }

    public String getWaName() {
        try {
            return (String) fieldGetWaName.get(mInstance);
        } catch (Exception e) {
            XposedBridge.log(e);
        }
        return null;
    }

    public static WaContactWpp getWaContactFromJid(FMessageWpp.UserJid userJid) {
        if (mInstanceGetWaContact == null) {
            XposedBridge.log("WAE: WaContactWpp: mInstanceGetWaContact is NULL. ContactManager not initialized?");
            return null;
        }
        try {
            Object contact = null;
            if (userJid.userJid != null) {
                contact = getWaContactMethod.invoke(mInstanceGetWaContact, userJid.userJid);
            }

            // Fallback to phoneJid if userJid lookup failed or userJid was null
            if (contact == null && userJid.phoneJid != null) {
                XposedBridge.log("WAE: WaContactWpp: userJid lookup failed, trying phoneJid");
                contact = getWaContactMethod.invoke(mInstanceGetWaContact, userJid.phoneJid);
            }

            if (contact != null) {
                return new WaContactWpp(contact);
            } else {
                XposedBridge.log("WAE: WaContactWpp: Contact lookup returned null for " + userJid);
            }
        } catch (Exception e) {
            XposedBridge.log(e);
        }
        return null;
    }

    public File getProfilePhoto() {
        try {
            return (File) getProfilePhoto.invoke(mInstanceGetProfilePhoto, mInstance);
        } catch (Exception e) {
            XposedBridge.log(e);
        }
        return null;
    }

}

File: app/src/main/java/com/wmods/wppenhacer/xposed/core/components/WaContactWpp.kt
package com.wmods.wppenhacer.xposed.core.components

import com.wmods.wppenhacer.xposed.core.devkit.Unobfuscator
import com.wmods.wppenhacer.xposed.utils.ReflectionUtils
import de.robv.android.xposed.XC_MethodHook
import de.robv.android.xposed.XposedBridge
import org.luckypray.dexkit.query.enums.StringMatchType
import java.io.InputStream
import java.lang.reflect.Field
import java.lang.reflect.Method

class WaContactWpp(instance: Any?) {

    val mInstance: Any

    init {
        if (instance == null) {
            throw RuntimeException("instance cannot be null")
        }
        if (!TYPE.isInstance(instance)) {
            throw RuntimeException("object is not a WaContactWpp")
        }
        this.mInstance = instance
    }

    companion object {


        lateinit var TYPE: Class<*>

        // Fields
        private var fieldContactData: Field? = null
        private var fieldUserJid: Field? = null
        private var fieldGetWaName: Field? = null
        private var fieldDataGetDisplayName: Field? = null
        private var fieldWaContactInData: Field? = null

        // Methods
        private var getWaContactMethod: Method? = null
        private var getProfilePhoto: Method? = null

        // Instances
        private var mInstanceGetWaContact: Any? = null
        private var mInstanceGetProfilePhoto: Any? = null

        @JvmStatic
        fun initialize(classLoader: ClassLoader) {
            try {
                TYPE = Unobfuscator.loadWaContactClass(classLoader)
                val classPhoneUserJid = Unobfuscator.findFirstClassUsingName(
                    classLoader,
                    StringMatchType.EndsWith,
                    "jid.PhoneUserJid"
                )
                val classJid = Unobfuscator.findFirstClassUsingName(
                    classLoader,
                    StringMatchType.EndsWith,
                    "jid.Jid"
                )

                val phoneUserJid = ReflectionUtils.getFieldByExtendType(TYPE, classPhoneUserJid)
                if (phoneUserJid == null) {
                    val contactDataClass = Unobfuscator.loadWaContactDataClass(classLoader)
                    fieldContactData = ReflectionUtils.getFieldByType(TYPE, contactDataClass)
                    fieldUserJid = ReflectionUtils.getFieldByExtendType(contactDataClass, classJid)
                } else {
                    fieldUserJid = ReflectionUtils.getFieldByExtendType(TYPE, classJid)
                }

                fieldDataGetDisplayName =
                    Unobfuscator.loadWaContactDataDisplayNameMethod(classLoader)
                fieldGetWaName = Unobfuscator.loadWaContactGetWaNameField(classLoader)

                getWaContactMethod = Unobfuscator.loadGetWaContactMethod(classLoader)
                getWaContactMethod?.let { method ->
                    XposedBridge.hookAllConstructors(
                        method.declaringClass,
                        object : XC_MethodHook() {
                            override fun afterHookedMethod(param: MethodHookParam) {
                                mInstanceGetWaContact = param.thisObject
                            }
                        })
                }

                getProfilePhoto = Unobfuscator.loadGetProfilePhotoMethod(classLoader)
                getProfilePhoto?.let { method ->
                    XposedBridge.hookAllConstructors(
                        method.declaringClass,
                        object : XC_MethodHook() {
                            override fun afterHookedMethod(param: MethodHookParam) {
                                mInstanceGetProfilePhoto = param.thisObject
                            }
                        })
                }

                fieldWaContactInData = fieldContactData!!.type.declaredFields.first {
                    it.type == TYPE
                }


            } catch (e: Exception) {
                XposedBridge.log(e)
            }
        }

        @JvmStatic
        fun getWaContactFromJid(userJid: FMessageWpp.UserJid): WaContactWpp? {
            return try {
                val instance = mInstanceGetWaContact
                val method = getWaContactMethod
                if (instance != null && method != null) {
                    var jid = userJid.userJid
                    if (!method.parameterTypes[0].isInstance(jid)) {
                        jid = FMessageWpp.UserJid.forceConverter(jid).userJid
                    }
                    val result = method.invoke(instance, jid)
                    result?.let { WaContactWpp(it) }
                } else null
            } catch (e: Exception) {
                XposedBridge.log(e)
                null
            }
        }
    }

    fun getObject(): Any = mInstance

    val userJid: FMessageWpp.UserJid by lazy {
        try {
            val userJidObj = fieldUserJid?.get(waContactData)
            userJidObj?.let { FMessageWpp.UserJid(it) } ?: FMessageWpp.UserJid()
        } catch (e: Exception) {
            XposedBridge.log(e)
            FMessageWpp.UserJid()
        }
    }

    val displayName: String?
        get() {
            return try {
                fieldDataGetDisplayName?.get(waContactData) as? String
            } catch (e: Exception) {
                XposedBridge.log(e)
                null
            }
        }

    val waName: String?
        get() {
            return try {
                if (fieldContactData!!.type.isAssignableFrom(fieldGetWaName!!.declaringClass)) {
                    return fieldGetWaName?.get(waContactData) as? String
                }
                fieldGetWaName?.get(mInstance) as? String
            } catch (e: Exception) {
                XposedBridge.log(e)
                null
            }
        }


    val waContactData: Any? by lazy {
        try {
            fieldContactData?.get(mInstance)
        } catch (e: Exception) {
            XposedBridge.log(e)
            null
        }
    }

    fun getProfilePhoto(fullImage: Boolean): InputStream? {
        return try {
            val instance = mInstanceGetProfilePhoto
            val method = getProfilePhoto
            if (instance != null && method != null) {
                method.invoke(instance, mInstance, fullImage) as? InputStream
            } else null
        } catch (e: Exception) {
            XposedBridge.log(e)
            null
        }
    }


}